File: src/client/AM.PA.MacOs/PersonalAnalytics/Trackers/Activity Tracker/Visualizations/DayProgramsUsedPieChart.swift
//
//  DayProgramsUsedPieChart.swift
//  PersonalAnalytics
//
//  Created by Chris Satterfield on 2017-05-29.
//
//  Adapted from Windows version created by André Meyer


import Foundation

class DayProgamsUsedPieChart: Visualization {
    
    let Size:String
    let color = AppConstants.retrospectiveColor
    var title: String
    let sql: ActivitySQLController
    var _type: [String] = [VisConstants.Day]
    let _minTimeWorked = 0.0
    
    required init() throws{
        Size = "Square"
        sql = try ActivitySQLController()
        title = "Top Programs Used"
    }
    
    func getHoursWorked(dict: Dictionary<String, Double> ) -> Double{
        var sum: Double = 0
        for (_, value) in dict{
            sum += value
        }
        return sum
    }
    
    func getHtml(_ _date: Date, type: String) -> String {
        
        if(!_type.contains(type)){
            print("return")
            return ""
        }
        var html = ""
        /////////////////////
        // fetch data sets
        /////////////////////
        //var chartQueryPieChartData = new Dictionary<string, long>();
        var chartQueryResultsLocal: [String:Double]
        chartQueryResultsLocal = sql.GetActivityPieChartData(date: _date)
        
        
        // merge with remote data if necessary //TODO: REMOTE DATA
        //chartQueryPieChartData = RemoteDataHandler.VisualizeWithRemoteData()
        //    ? RemoteDataHandler.MergeActivityData(chartQueryResultsLocal, Queries.GetActivityPieChartData(_date))
        //    : chartQueryResultsLocal;
        
        /////////////////////
        // data cleaning
        /////////////////////
        //// remove IDLE (doesn't belong to activity on computer)
        //if (chartQueryResultsLocal.ContainsKey(Dict.Idle))
        //    chartQueryResultsLocal.Remove(Dict.Idle);
        
        // calculate total active time
        let totalHoursWorked = getHoursWorked(dict: chartQueryResultsLocal)
        print("total hours:",totalHoursWorked)
        // check if we have enough data
        if (chartQueryResultsLocal.count == 0 || totalHoursWorked < _minTimeWorked)
        {
            html += VisHelper.NotEnoughData()
            return html;
        }
    
        PrepareDataForVisualization(chartQueryResultsLocal)
        
        
        /////////////////////
        // HTML
        /////////////////////
        html += "<p style='text-align: center;'>Total hours worked on your computer: <strong>" + String(totalHoursWorked) + "</strong>.</p>";
        html += "<div id='" + VisHelper.CreateChartHtmlTitle(title: title) + "' style='height:75%;'  align='center'></div>"
        
        
        /////////////////////
        // JS
        /////////////////////
        var columns = ""
        for (program, time) in chartQueryResultsLocal{
            columns += "['\(program)', \(time)], "
        }
        
        let data = "columns: [ " + columns + "], type: 'pie'"
        
        html += "<script type='text/javascript'>";
        
        let timeType = "h"
        
        html += "var " + VisHelper.CreateChartHtmlTitle(title: title) + " = c3.generate({ bindto: '#" + VisHelper.CreateChartHtmlTitle(title: title) + "', data: { " + data + "}, pie: { label: { format: function (value, ratio, id) { return value + '" + timeType + "';}}}, padding: { top: 0, right: 0, bottom: 0, left: 0 }, legend: { show: true, position: 'bottom' }});"
        html += "</script>"
        
        return html
    }
    
    /// <summary>
    /// Adds all items in the list (in case more than 10) to an Other group
    /// </summary>
    /// <param name="chartQueryResultsLocal"></param>
    func PrepareDataForVisualization(_ chartQueryResultsLocal: Dictionary<String,Double>){
          /*  //if (chartQueryResultsLocal.Count >= _maxNumberOfPrograms)
            //{
            // summarize small parts of work
            var totalHoursWorked = getHoursWorked(chartQueryResultsLocal)

        
        // remove OTHER if it is too small
            if (chartQueryResultsLocal.ContainsKey(Dict.Other) && chartQueryResultsLocal[Dict.Other] <= small)
                keysToRemove.Add(Dict.Other);
                
            foreach (var key in keysToRemove)
            {
                chartQueryResultsLocal.Remove(key);
            }*/
                
            title = "Top \(String(chartQueryResultsLocal.count)) Programs Used"
    }
    
    
}

File: src/windows/AM.PA.MonitoringTool/WindowsActivityTracker/Visualizations/DayProgramsUsedPieChart.cs
// Created by André Meyer (ameyer@ifi.uzh.ch) from the University of Zurich
// Created: 2015-10-20
// 
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using Shared;
using Shared.Helpers;
using WindowsActivityTracker.Data;
using System.Globalization;

namespace WindowsActivityTracker.Visualizations
{
    internal class DayProgramsUsedPieChart : BaseVisualization, IVisualization
    {
        private readonly DateTimeOffset _date;
        //private const int _maxNumberOfPrograms = 10;
        private const double _minTimeWorked = 0.3; // in hours

        public DayProgramsUsedPieChart(DateTimeOffset date)
        {
            this._date = date;

            Title = "Top Programs Used"; //hint; overwritten below
            IsEnabled = true; //todo: handle by user
            Order = 1; //todo: handle by user
            Size = VisSize.Square;
            Type = VisType.Day;
        }

        public override string GetHtml()
        {
            var html = string.Empty;

            /////////////////////
            // fetch data sets
            /////////////////////
            //var chartQueryPieChartData = new Dictionary<string, long>();
            var chartQueryResultsLocal = Queries.GetActivityPieChartData(_date);

            // merge with remote data if necessary //TODO: REMOTE DATA
            //chartQueryPieChartData = RemoteDataHandler.VisualizeWithRemoteData()
            //    ? RemoteDataHandler.MergeActivityData(chartQueryResultsLocal, Queries.GetActivityPieChartData(_date))
            //    : chartQueryResultsLocal;

            /////////////////////
            // data cleaning
            /////////////////////
            //// remove IDLE (doesn't belong to activity on computer)
            //if (chartQueryResultsLocal.ContainsKey(Dict.Idle))
            //    chartQueryResultsLocal.Remove(Dict.Idle);

            // calculate total active time
            var totalHoursWorked = chartQueryResultsLocal.Sum(x => x.Value);

            // check if we have enough data
            if (chartQueryResultsLocal.Count == 0 || totalHoursWorked < _minTimeWorked)
            {
                html += VisHelper.NotEnoughData(Dict.NotEnoughData);
                return html;
            }

            PrepareDataForVisualization(chartQueryResultsLocal);


            /////////////////////
            // HTML
            /////////////////////
            html += "<p style='text-align: center;'>Total hours worked on your computer: <strong>" + Math.Round(totalHoursWorked, 1) + "</strong>.</p>";
            html += "<div id='" + VisHelper.CreateChartHtmlTitle(Title) + "' style='height:75%;'  align='center'></div>";


            /////////////////////
            // JS
            /////////////////////
            var columns = string.Empty;
            foreach (var program in chartQueryResultsLocal)
            {
                columns += string.Format(CultureInfo.InvariantCulture, "['{0}', {1}], ", program.Key, Math.Round(program.Value, 1));
            }

            var data = "columns: [ " + columns + "], type: 'pie'";

            html += "<script type='text/javascript'>";
            html += "var " + VisHelper.CreateChartHtmlTitle(Title) + " = c3.generate({ bindto: '#" + VisHelper.CreateChartHtmlTitle(Title) + "', data: { " + data + "}, pie: { label: { format: function (value, ratio, id) { return value + 'h';}}}, padding: { top: 0, right: 0, bottom: 0, left: 0 }, legend: { show: true, position: 'bottom' }});";
            html += "</script>";

            return html;
        }

        /// <summary>
        /// Adds all items in the list (in case more than 10) to an Other group
        /// </summary>
        /// <param name="chartQueryResultsLocal"></param>
        private void PrepareDataForVisualization(Dictionary<string, double> chartQueryResultsLocal)
        {
            try
            {
                //if (chartQueryResultsLocal.Count >= _maxNumberOfPrograms)
                //{
                    // summarize small parts of work
                    var totalHoursWorked = chartQueryResultsLocal.Sum(x => x.Value);
                    var small = (totalHoursWorked > 1.0) ? totalHoursWorked * 0.05 : totalHoursWorked * 0.1; // put to 'other' either 5% or 10% of the processes
                    var keysToRemove = new List<string>();

                    foreach (var item in chartQueryResultsLocal.ToList())
                    {
                        if (item.Value < small)
                        {
                            if (chartQueryResultsLocal.ContainsKey(Dict.Other))
                            {
                                chartQueryResultsLocal[Dict.Other] += item.Value;
                            }
                            else
                            {
                                chartQueryResultsLocal.Add(Dict.Other, item.Value);
                            }
                            keysToRemove.Add(item.Key);
                        }
                    }

                    // remove OTHER if it is too small
                    if (chartQueryResultsLocal.ContainsKey(Dict.Other) && chartQueryResultsLocal[Dict.Other] <= small)
                        keysToRemove.Add(Dict.Other);

                    foreach (var key in keysToRemove)
                    {
                        chartQueryResultsLocal.Remove(key);
                    }
                //}

                Title = string.Format(CultureInfo.InvariantCulture, "Top {0} Programs Used", chartQueryResultsLocal.Count);
            }
            catch (Exception e)
            {
                Logger.WriteToLogFile(e);
            }
        }
    }
}
